### ISSUE TITLE
file: mode=2750 != {mode: 2750}

### ISSUE THREAD

**Issue by groks** [ansible/ansible#5409]
Some kind of confusion over octal numbers?

```
- name: this works
  file: path=/tmp/ok state=directory mode=2750

- name: this fails
  file: {path: /tmp/nok, state: directory, mode: 2750}

- name: this also fails
  file:
    path: /tmp/nope
    state: directory
    mode: 2750
```

The permissions look like this:

```
drwxr-s---.    /tmp/ok
d-wSrwxrwT.    /tmp/nok
d-wSrwxrwT.    /tmp/nope
```

Only the first one is correct.

**Comment by sivel** [comment]
This is due to a difference in yaml processing and the shlex processing used for `key=value` arguments, and not using proper python octal references.

When using `key=value`, such as `mode=2750`, the value ends up as a string, as opposed to a numeric integer.

When using `key: value` such as `mode: 2750`, the value ends up as a base 10 integer.

Because `os.chmod` can only accept a numeric entry, the code does a `int(mode, 8)` when the value is a string to convert it to an integer specifying the original base as base 8 (octal).

`2750` is an invalid octal mode in python.  The correct octal mode in python is `02750`.  And the python yaml implementation will parse `mode: 02750` and convert the value automatically to a base 10 integer of `1512`.

So, in short you should always prefix an octal file mode with a `0`, whether using `mode=02750` or `mode: 02750`

Whether a change should be made in ansible I don't know.  I am thinking it shouldn't, as we cannot be sure what the user really wanted.  If 2750 is provided, we cannot necessarily assume that it is octal, because base 10 integers are also valid.

**Comment by mpdehaan** [comment]
mode: yep, decimal 2750 is an decimal integer

You should quote the modes or use the octal prefix.

Ansible prefers key=value anyway, which is the way most people do it, and limits possible confusion.
